Smash the state

Gastove opened the pull request:
Well.
#### Holy Shit

So this got a little intense. The goal is the resolution of #37, which will eventually be achieved by making the `Doctopus` stack approximately stateless. From there, things got a little hinky. 

The first thing I realized was that a `Head` and a `Tentacle` need a way to be mapped to each other -- but neither of those should be in charge of it. So, I started in on a join table (`head_tentacle_mappings`). Then I needed Korma to do lookups in the join table, and I wanted unit tests to make sure those were right. As long as I was there, I wanted unit tests for the rest of our DB interactions -- which lead me to some bug and function-fixes.

As this stands, it's... getting there. Some changes haven't been pushed yet (apologies), and this is definitely WIP. All the same: comments gratefully welcomed.

Gastove committed:
Rename db/core.clj to db.clj

Gastove committed:
Propagate rename

Gastove committed:
formatting

Gastove committed:
Start wiring up a join table

Gastove committed:
Schema for the join table; tweak the bootstrapping function

Gastove committed:
Bring in the db ns; docstrang

Gastove committed:
Merge branch 'master' into smash-the-state

Gastove committed:
This... _should_ wire up our join table correct?

Gastove committed:
Bring in clojure.string; let my ide sort ns imports; formatting

Gastove committed:
Correct schema syntax for Postgres (MySql, derp)

Gastove committed:
We... might need this later? It's nice now. Adding a function that tries
a sql statement and logs it _and the full stacktrace_ if something goes pear-shaped

Gastove committed:
Typo

Gastove committed:
Add some data-mocking test utils

Gastove committed:
Make delete functions work by passing a head, not just the name of a head

Gastove committed:
In order to support different database configs (say, "main" versus
"test"), refactor this to build a config map by passing in the name of
the database to configure.

Gastove committed:
Cascade deletes to the mapping table. This 1) allows deletes of heads at
all (good!) and means that mappings cant point to null heads (also positive!)

Gastove committed:
Pass around a db name; build configs as needed

Gastove committed:
Update the table-created call to use a db name

Gastove committed:
Expand the test utilities to support mocking out heads and
tentacles. This is.... on the clunky side at the moment (second arg
unused); think of a better way to do this.

Gastove committed:
Start unit testing our CRUD layer :D

Add a testing fixture to set the correct database, bootstrap it,  and perform a teardown
at the end.

Gastove committed:
Take out head name

Gastove committed:
Update travis.yml for postgres

Gastove committed:
Add a DB for travis

Gastove committed:
Fully flesh out CRUD/lookup API unit tests

Gastove committed:
Add a compound primary key to avoid duplicate mappings

Gastove committed:
Fix until unit tests pass :D

Gastove committed:
forgot a damn :

Gastove committed:
lergs

Gastove committed:
Strip out a head's state. It now loads all its tentacles, on demand,
from the db, every time.

Gastove committed:
First pass: rip the state out of Doctopus. It should now load fresh data
every time

Gastove committed:
Change the name of save-<foo>! to create-<foo>!; adds a new save-foo!
method that checks for the existence of foo, then dispatches to either
create or update.

jarofghosts commented on resources/configuration.edn:
```diff
@@ -7,5 +7,8 @@
                       "travis" {:heads-dir "resources/self/heads"
                                 :permanent-fs-root "~/"
-                                :ip 8080}}
+                                :ip 8080
```
why is this `:ip` vs. `:port` ?

Gastove replied:
b/c derp ;-P

Gastove replied:
Bizarre but true, this has been passing Travis builds? idek

jarofghosts commented on src/doctopus/db.clj:
```diff
@@ -0,0 +1,199 @@
+                            :tentacles.source_control
+                            :tentacles.entry_point))
+         :tentacles_by_head)
+  ;; These don't seem to have any effect and I have no idea why.
```
:(

Gastove replied:
Yeah, that part is fucky. I'm still trying to figure out if I'm doing something wrong, or if it's a bug in Korma, or.... space.... goblins?

jarofghosts replied:
evidently [this or something like it was a bug at one time](https://github.com/korma/Korma/issues/215), i wonder if this is a regression?

jarofghosts replied:
i am going to see if i can reproduce

jarofghosts replied:
also, you should be able to drop the `prepare` since you won't be inserting through this entity

Gastove replied:
Aw man, that's awesome -- thank you for digging in to that! I'm sure curious what's going on -- my own nincompoopery, or a bug, or what.

jarofghosts commented on src/doctopus/db.clj:
```diff
@@ -0,0 +1,199 @@
+  [tentacle]
+  (log/warn "Deleting tentacle:" tentacle "!")
+  (delete tentacles
+          (where {:name (:name tentacle)})))
```
!m

jarofghosts replied:
(i'm not sure why i did this)

Gastove committed:
Refactor DB fixture to its own NS

Gastove committed:
Remove state

Gastove committed:
Root no longer needed

Gastove committed:
derp

Gastove committed:
Protip: Korma can't serialize keywords

Gastove committed:
Add a sc field to the mock

Gastove committed:
Switch to using Korma (withs) such that returns are now much nicer

Gastove committed:
Nicer var name

Gastove committed:
This should stop killing everything

Gastove committed:
Ooookay. This, I think, rips out or updates all the parts of these tests
that currently melt. Last thing to do is to write new, good tests and
we're set to go. I think.

jarofghosts commented on resources/test/heads/test/doctopus-test.edn:
```diff
@@ -1,6 +1,6 @@
  :html-commands ["lein marg -f index.html"]
  :name "doctopus-test"
- :source-control :git
+ :source-control "git"
```
i had considered possibly doing the `prepare` / `transform` dance to make these keywords, thoughts?

Gastove replied:
I am A+ on doing that eventually. This was one of the subtle bugs I had to
dig up last night -- and right now, since git is the only DVCS we support,
we aren't actually using that field. I'd say, let's make an issue to
`prepare`/`transform` that, maybe also to wire up, say, mercurial?

On Tue, May 5, 2015 at 10:32 AM, jesse keane notifications@github.com
wrote:

> In resources/test/heads/test/doctopus-test.edn
> https://github.com/Gastove/doctopus/pull/45#discussion_r29692295:
> 
> > @@ -1,6 +1,6 @@
> >  {:output-root "docs"
> >   :html-commands ["lein marg -f index.html"]
> >   :name "doctopus-test"
> > - :source-control :git
> > - :source-control "git"
> 
> i had considered possibly doing the prepare / transform dance to make
> these keywords, thoughts?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Gastove/doctopus/pull/45/files#r29692295.

Gastove committed:
Change the fixture to do a full pave-and-re-bootstrap before and after
ever set of tests.

Gastove committed:
Maps don't have guaranteed order -- which I knew, but forgot. Without
guaranteed order, there's a decent chance this will try to create the
mappings table before one or both parent tables. Glee.

Gastove committed:
Add a loader function to be part of bootstrapping that creates one head
and one tentacle

jarofghosts commented:
:+1::+1::+1: :+1::+1::+1:
:+1::+1::+1: :+1::+1::+1:
:+1::+1::+1: :+1::+1::+1:

Gastove committed:
Of course, using the current tests against the code on master
explodes. This fixes that, BUT, once this PR is merged, this change
should totally be undone

Gastove committed:
Fix DB tests to work with the new bootstrapping

Gastove committed:
Fix head tests

Gastove committed:
Fix database tests

Gastove committed:
truncate! needs to not be private any more
